gobblin-core/src/main/java/gobblin/writer/FsDataWriter.java

```diff
@@ -0,0 +1,76 @@
+/* (c) 2015 NerdWallet All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.writer;
+
+import java.io.IOException;
+import java.net.URI;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.configuration.State;
+import gobblin.util.ForkOperatorUtils;
+import gobblin.util.HadoopUtils;
+import gobblin.util.WriterUtils;
+
+/**
+ * An implementation of {@link DataWriter} does the work of setting the output/staging dir
+ * and creating the FileSystem instance.
+ *
+ * @author akshay@nerdwallet.com
+ */
+public abstract class FsDataWriter<D> implements DataWriter<D> {
+  private static final Logger LOG = LoggerFactory.getLogger(FsDataWriter.class);
+
+  protected final FileSystem fs;
+  protected final Path stagingFile;
+  protected final Path outputFile;
+  protected final State properties;
+
+  public FsDataWriter(State properties, String fileName, int numBranches, int branchId) throws IOException {
+    this.properties = properties;
+    // initialize file system
+    String uri = properties.getProp(
+            ForkOperatorUtils.getPropertyNameForBranch(ConfigurationKeys.WRITER_FILE_SYSTEM_URI, numBranches, branchId),
+            ConfigurationKeys.LOCAL_FS_URI
+    );
+    Configuration conf = new Configuration();
+    // Add all job configuration properties so they are picked up by Hadoop
+    for (String key : properties.getPropertyNames()) {
+      conf.set(key, properties.getProp(key));
+    }
+    this.fs = FileSystem.get(URI.create(uri), conf);
+
+    // initialize staging/output dir
+    this.stagingFile = new Path(WriterUtils.getWriterStagingDir(properties, numBranches, branchId), fileName);
+    this.outputFile = new Path(WriterUtils.getWriterOutputDir(properties, numBranches, branchId), fileName);
+    properties.setProp(ForkOperatorUtils.getPropertyNameForBranch(ConfigurationKeys.WRITER_FINAL_OUTPUT_PATH, branchId),
```

Comment:
Replacing `WRITER_FINAL_OUTPUT_PATH` with `WRITER_FILE_PATH` doesn't work because `WRITER_FINAL_OUTPUT_PATH = WRITER_OUTPUT_DIR + WRITER_FILE_PATH`. Given that, I don't think you need `WRITER_FINAL_OUTPUT_PATH`, you can always set `WRITER_OUTPUT_DIR` and `WRITER_FILE_PATH` if you want to change where the writer outputs. @liyinan926 what do you think?